src/outputs/restart.cpp

```diff
@@ -17,30 +17,582 @@
+
+  // open blocks tab
+  hid_t gBlocks = H5Gcreate(file, "/Blocks", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+
+  // write Xmin[ndim] for blocks
+  {
+    Real *tmpData = new Real[num_blocks_local * 3];
+    local_count[0] = num_blocks_local;
+    global_count[0] = max_blocks_global;
+    pmb = pm->pblock;
+    int i = 0;
+    while (pmb != nullptr) {
+      auto xmin = pmb->coords.GetXmin();
+      tmpData[i] = xmin[0];
+      i++;
+      if (pm->ndim > 1) {
+        tmpData[i] = xmin[1];
+        i++;
+      }
+      if (pm->ndim > 2) {
+        tmpData[i] = xmin[2];
+        i++;
+      }
+      pmb = pmb->next;
+    }
+    local_count[1] = global_count[1] = pm->ndim;
+    WRITEH5SLABDOUBLE("xmin", tmpData, gBlocks, local_start, local_count, global_count,
+                      property_list);
+    delete[] tmpData;
+  }
+
+  // write Block ID
+  {
+    // LOC.lx1,2,3
+    hsize_t n;
+    int i;
+
+    n = 3;
+    auto *tmpLoc = new int64_t[num_blocks_local * n];
+    local_count[1] = global_count[1] = n;
+    local_count[0] = num_blocks_local;
+    global_count[0] = max_blocks_global;
+    pmb = pm->pblock;
+    i = 0;
+    while (pmb != nullptr) {
+      tmpLoc[i++] = pmb->loc.lx1;
+      tmpLoc[i++] = pmb->loc.lx2;
+      tmpLoc[i++] = pmb->loc.lx3;
+      pmb = pmb->next;
+    }
+    WRITEH5SLABI64("loc.lx123", tmpLoc, gBlocks, local_start, local_count, global_count,
+                   property_list);
+    delete[] tmpLoc;
+
+    // (LOC.)level, GID, LID, cnghost, gflag
+    n = 5;
+    auto *tmpID = new int[num_blocks_local * n];
+    local_count[1] = global_count[1] = n;
+    local_count[0] = num_blocks_local;
+    global_count[0] = max_blocks_global;
+    pmb = pm->pblock;
+    i = 0;
+    while (pmb != nullptr) {
+      tmpID[i++] = pmb->loc.level;
+      tmpID[i++] = pmb->gid;
+      tmpID[i++] = pmb->lid;
+      tmpID[i++] = pmb->cnghost;
+      tmpID[i++] = pmb->gflag;
+      pmb = pmb->next;
+    }
+    WRITEH5SLABI32("loc.level-gid-lid-cnghost-gflag", tmpID, gBlocks, local_start,
+                   local_count, global_count, property_list);
+    delete[] tmpID;
+  }
+
+  // close locations tab
+  H5Gclose(gBlocks);
+
+  // write variables
+
+  // write variables
+  // create persistent spaces
+  local_count[1] = nx3;
+  local_count[2] = nx2;
+  local_count[3] = nx1;
+  local_count[4] = 1;
+
+  global_count[1] = nx3;
+  global_count[2] = nx2;
+  global_count[3] = nx1;
+  global_count[4] = 1;
+
+  hid_t local_DSpace = H5Screate_simple(5, local_count, NULL);
+  hid_t global_DSpace = H5Screate_simple(5, global_count, NULL);
+
+  // while we could do this as n variables and load all variables for
+  // a block at one time, this is memory-expensive.  I think it is
+  // well worth the multiple iterations through the blocks to load up
+  // one variable at a time.  Besides most of the time will be spent
+  // writing the HDF5 file to disk anyway...
+  // If I'm wrong about this, we can always rewrite this later.
+  // Sriram
```

Comment:
When we're running at scale, this will probably deserve some profiling. But I agree, we can rethink it later.

Reply:
👍